x-hooks/use-authentication.tsx

```diff
@@ -0,0 +1,209 @@
+  const {getUserOf, getUserWith} = useApi();
+
+  const [lastUrl,] = useState(new WinStorage('lastUrlBeforeGHConnect', 0, 'sessionStorage'));
+  const [balance,] = useState(new WinStorage('lastUrlBeforeGHConnect', 1000, 'sessionStorage'));
+
+  const URL_BASE = typeof window !== "undefined" ? `${window.location.protocol}//${ window.location.host}` : "";
+
+  function disconnectGithub() {
+    return signOut({redirect: false});
+  }
+
+  function disconnectWallet() {
+
+    if (!state.currentUser?.walletAddress)
+      return;
+
+    signOut({callbackUrl: `${URL_BASE}/${state.Service.network.lastVisited}`})
+      .then(() => {
+        dispatch(changeCurrentUser.update({handle: state.currentUser?.handle, walletAddress: ''}));
+      });
+  }
+
+  function connectWallet() {
+    if (!state.Service?.active)
+      return;
+
+    connect();
+  }
+
+  function updateWalletAddress() {
+    if (state.spinners?.wallet)
+      return;
+
+    if (!state.currentUser?.connected)
+      return;
+
+    console.log(`updating wallet`)
+
+    dispatch(changeWalletSpinnerTo(true));
+
+    state.Service.active.getAddress()
+      .then(address => {
+        if (address !== state.currentUser?.walletAddress)
+          dispatch(changeCurrentUserWallet(address))
+      })
+      .catch(e => {
+        console.error(`Error getting address`, e);
+      })
+      .finally(() => {
+        dispatch(changeWalletSpinnerTo(false));
+      })
+
+  }
+
+  function connectGithub() {
+    console.debug(`connectGithub`, state.currentUser)
+
+    if (!state.currentUser?.walletAddress)
+      return;
+
+    getUserOf(state.currentUser?.walletAddress)
+      .then((user) => {
+        if (!user?.githubLogin && !asPath.includes(`connect-account`)) {
+          disconnectGithub()
+          push(`/connect-account`);
+          return false;
+        }
+        return true;
+      })
+      .then(signedIn => {
+        if (!signedIn)
+          return;
+
+        lastUrl.value = asPath;
+
+        return signedIn ? signIn('github', {callbackUrl: `${URL_BASE}${asPath}`}) : null;
+      })
+  }
+
+  function validateGhAndWallet() {
+    if (!state.currentUser?.walletAddress || !(session?.data?.user as any)?.login || state.spinners?.matching)
+      return;
+
+    dispatch(changeSpinners.update({matching: true}));
+
+    const userLogin = (session.data.user as any).login;
+    const walletAddress = state.currentUser.walletAddress.toLowerCase();
+
+    getUserWith(userLogin)
+      .then(user => {
+        if (!user.githubLogin && state.currentUser?.match !== undefined)
+          dispatch(changeCurrentUserMatch(undefined));
+        else if (user.githubLogin && userLogin)
+          dispatch(changeCurrentUserMatch(userLogin === user.githubLogin &&
+            (walletAddress ? walletAddress === user.address : true)));
+      })
+      .finally(() => {
+        dispatch(changeSpinners.update({matching: false}));
+      })
+  }
+
+  function listenToAccountsChanged() {
```

Comment:
The way this functions is being used creates the following side effect: if the user didn't clicked on the "Connect Wallet" button yet and change the wallet on metamask it will automatically connect to the app. Is it the desired behavior ?